sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (C) 2021 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    net_gnrc_ipv6_auto_subnets Simple-Subnet auto-configuration
+ * @ingroup     net_gnrc
+ * @brief       Automatic configuration for cascading subnets
+ *
+ * About
+ * =====
+ *
+ * This module provides an automatic configuration for networks with a simple
+ * tree topology.
+ *
+ * If a sufficiently large IPv6 prefix is provided via Router Advertisements, a
+ * routing node with this module will automatically configure subnets from it for
+ * each downstream interface.
+ *
+ * There can only be a single routing node on each level of the network but an
+ * arbitrary number of leaf nodes.
+ *
+ * ![Example Topology](gnrc_ipv6_auto_subnets_simple.svg)
+ *
+ * The downstream network(s) get the reduced prefix via Router Advertisements and
+ * the process repeats until the bits of prefix are exhausted. (The smallest subnet
+ * must still have a /64 prefix.)
+ *
+ * The downstream router will send a router advertisement with only a
+ * Route Information Option included to the upstream network.
+ * The Route Information Option contains the prefix of the downstream network so that
+ * upstream hosts will no longer consider hosts in this subnet on-link but instead
+ * will use the downstream router to route to the new subnet.
+ *
+ * Usage
+ * =====
+ *
+ * Simply add the `gnrc_ipv6_auto_subnets` module to the nodes that should act as routers
+ * in the cascading network.
+ * The upstream network will be automatically chosen as the one that first receives a
+ * router advertisement.
+ *
+ * @{
+ *
+ * @file
+ * @author  Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#include "net/gnrc/ipv6/nib.h"
+#include "net/gnrc/ndp.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static char addr_str[IPV6_ADDR_MAX_STR_LEN];
+
+static void _init_sub_prefix(ipv6_addr_t *out,
+                             const ipv6_addr_t *prefix, uint8_t bits,
+                             uint8_t idx, uint8_t idx_bits)
+{
+    uint8_t bytes = bits / 8;
+    uint8_t rem   = bits % 8;
+    int8_t shift  = 8 - rem - idx_bits;
+
+    /* first copy old prefix */
+    memcpy(out, prefix, bytes);
+
+    /* make sure unused bits are cleared */
+    if (rem) {
+        uint8_t mask = 0xff << (8 - rem);
+        out->u8[bytes] = prefix->u8[bytes] & mask;
+    } else {
+        out->u8[bytes] = 0;
+    }
```

Comment:
Unless you are under timing constraints, I think this can be solved easier to understand (and I think also with smaller code footprint, as the functions is already used elsewhere in GNRC), with

```suggestion
    memset(out, 0, sizeof(*out));
    ipv6_addr_init_prefix(out, prefix, bits);
```